components/debounce/browser/debounce_navigation_throttle.cc

```diff
@@ -0,0 +1,163 @@
+/* Copyright 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/debounce/browser/debounce_navigation_throttle.h"
+
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "base/bind.h"
+#include "base/memory/weak_ptr.h"
+#include "base/threading/sequenced_task_runner_handle.h"
+#include "brave/components/debounce/browser/debounce_service.h"
+#include "content/public/browser/navigation_entry.h"
+#include "content/public/browser/navigation_handle.h"
+#include "content/public/browser/navigation_throttle.h"
+#include "content/public/browser/page_navigator.h"
+#include "content/public/browser/web_contents.h"
+#include "content/public/browser/web_contents_observer.h"
+#include "content/public/browser/web_contents_user_data.h"
+#include "ui/base/page_transition_types.h"
+#include "url/gurl.h"
+#include "url/origin.h"
+
+using content::NavigationHandle;
+using content::NavigationThrottle;
+using content::WebContents;
+
+namespace debounce {
+
+namespace {
+
+class DebounceTabHelper
+    : public content::WebContentsObserver,
+      public content::WebContentsUserData<DebounceTabHelper> {
+ public:
+  explicit DebounceTabHelper(content::WebContents* web_contents)
+      : content::WebContentsObserver(web_contents),
+        content::WebContentsUserData<DebounceTabHelper>(*web_contents) {}
+
+  ~DebounceTabHelper() override = default;
+
+  DebounceTabHelper(const DebounceTabHelper&) = delete;
+  DebounceTabHelper& operator=(const DebounceTabHelper&) = delete;
+
+  void AddToRedirectChain(const GURL& url) { redirects_.push_back(url.host()); }
+  void ClearRedirectChain() { redirects_.clear(); }
+  bool IsInRedirectChain(const GURL& url) {
```

Comment:
Given the Chromium bug we ran into for De-AMP with redirects, can we add a test for this?

Reply:
there is one

Reply:
https://github.com/brave/brave-core/pull/14185/files#diff-cf72a5bad0d8df5913155f95898766fd203cd17bcfad3930945a2fed35876781R268